src/core/umount.c

```diff
@@ -382,6 +384,82 @@ static bool nonunmountable_path(const char *path) {
                 || path_startswith(path, "/run/initramfs");
 }
 
+static int remount_with_timeout(MountPoint *m, char *options, int *n_failed, bool log_error) {
+        pid_t pid;
+        int r;
+
+        BLOCK_SIGNALS(SIGCHLD);
+
+        /* Due to the possiblity of a remount operation hanging, we
+         * fork a child process and set a timeout. If the timeout
+         * lapses, the assumption is that that particular remount
+         * failed. */
+        pid = fork();
+        if (pid < 0)
+                return log_error_errno(errno, "Failed to fork: %m");
+
+        if (pid == 0) {
+                log_info("Remounting '%s' read-only in with options '%s'.", m->path, options);
+
+                /* Start the mount operation here in the child */
+                r = mount(NULL, m->path, NULL, MS_REMOUNT|MS_RDONLY, options);
+                if (r < 0) {
+                        if (log_error)
+                                log_error_errno(errno, "Failed to remount '%s' read-only: %m", m->path);
+                }
+                _exit(r < 0 ? EXIT_FAILURE : EXIT_SUCCESS);
+        }
+
+        r = wait_for_terminate_with_timeout(pid, DEFAULT_TIMEOUT_USEC);
+        if (r < 0 && r == -ETIMEDOUT) {
+                log_error("Remounting '%s' - timed out, issuing SIGKILL to PID "PID_FMT".", m->path, pid);
```

Comment:
log_error_errno() please, after all we have an error to log

Reply:
I assume this is to log `errno` in the journal etc, and it's ok to not include "%m" in the message if we don't need to.

Reply:
yes, precisely.

(i mean, of course, the journal is not going to run this late, but still, we should follow the same logging rules everywhere, and include errno metadata wherever we have it, regardless where the code ends up being used. After all we shift code around every now and then and it might be reused in a different context and so on)